db/import.c

```diff
@@ -0,0 +1,2117 @@
+ *
+ * 4. The import database runs full recovery using the files it grabbed from the source.
+ *
+ * 5. The import database writes information about the import target table 
+ * to a file and then terminates.
+ *
+ * 6. The destination db reads the file written by the import database and uses this 
+ * information to copy the btree files associated with the target table from the 
+ * import database directory into its directory, copy the files to the replicants,
+ * and finally to overwrite the existing table with a schema change.
+ */
+
+#include <fcntl.h>
+#include <netdb.h>
+#include <pthread.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "fdb_fend.h"
+#include "bdb_api.h"
+#include "bdb_schemachange.h"
+#include "comdb2.h"
+#include "comdb2_appsock.h"
+#include "crc32c.h"
+#include "importdata.pb-c.h"
+#include "intern_strings.h"
+#include "logmsg.h"
+#include "sc_callbacks.h"
+#include "sc_global.h"
+#include "str0.h"
+#include "version_util.h"
+
+#define BULK_IMPORT_MIN_SUPPORTED_VERSION "8.1.0"
+
+#define __import_logmsg(lvl, msg, ...) \
+    logmsg(lvl, "[IMPORT] %s: " msg, __func__, ## __VA_ARGS__)
+
+static int value_on_off(const char *value) {
+    if (strcasecmp("on", value) == 0) {
+        return 1;
+    } else if (strcasecmp("off", value) == 0) {
+        return 0;
+    } else if (strcasecmp("no", value) == 0) {
+        return 0;
+    } else if (strcasecmp("yes", value) == 0) {
+        return 1;
+    }
+    return atoi(value);
+}
+
+extern char *comdb2_get_tmp_dir();
+extern int comdb2_get_tmp_dir_args(char *tmp_dir, ssize_t sz, const char *basedir, const char *envname, int nonames);
+extern void get_txndir_args(char *txndir, size_t sz_txndir, const char *dbdir, const char *envname, int nonames);
+
+extern tran_type *curtran_gettran(void);
+extern void curtran_puttran(tran_type *tran);
+
+extern int gbl_import_mode;
+extern char *gbl_import_table;
+extern char *gbl_file_copier;
+extern char gbl_dbname[MAX_DBNAME_LENGTH];
+extern char *gbl_dbdir;
+
+extern int should_ignore_btree(const char *filename,
+                               int (*should_ignore_table)(const char *),
+                               int should_ignore_queues,
+                               int name_boundary_exists);
+
+static uint64_t gbl_import_id = 0;
+pthread_mutex_t import_id_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+/* Constants */
+#define FILENAMELEN 100
+
+/* inputs can't be null and ending can't be emptystring */
+static int str_has_ending(const char * const s_start, const char * const e_start) {
+    const ssize_t s_len = strlen(s_start);
+    const ssize_t e_len = strlen(e_start);
+    if (e_len > s_len) { return 0; }
+
+    const char * e_last = e_start + e_len;
+    const char * s_last = s_start + s_len;
+
+    // if last char is neq then elast < estart
+    while ((*e_last-- == *s_last--) && (e_last >= e_start)) {}
```

Comment:
This is a bit too complicated.  It seems to also check a byte at the beginning of a terminating string, which might not be allocated.  May I suggest a strcmp for last e_len bytes?

Reply:
Thanks for pointing this out Dorin--I agree that it's overly-complicated.
I re-implemented the function using `strcmp()` [here](https://github.com/bloomberg/comdb2/pull/4473/commits/ca8ac6f27d26a1193d8d48f44ab1ee7762a2dcb4)